### ISSUE TITLE
Revisit Project Structure

### ISSUE THREAD

**Issue by danielpacak** [aquasecurity/tracee#1217]
## Abstract 

1. The structure of tracee is quite complex. (Though it’s very subjective.) 
2. We maintain three separate Go modules in the same Git repository even though we always release them altogether. This demonstrates that `tracee-ebpf` and `tracee-rules` executables have the same release lifecycle. 
3. Navigating between different Go modules that have their own Makefiles and Dockerfile builders is a bit overwhelming compared to the number of build artifacts that are produced. (We have quite some code and dependency duplication in Makefiles and Dockerfiles as well.) 
4. Adding a CI workflow that integration tests both tracee-ebpf and tracee-rules may be challenging as we even separated GitHub Actions workflows by Go modules. 

### TL;DR; 

Use Go project structure that sets explicit boundaries for `tracee-ebpf`, `tracee-rules` executables, and Go code that can be used as a library in a custom downstream integration. In other words, `tracee-ebpf` and `tracee-rules` are considered default command line interfaces, which handle CLI flags and arguments parsing and are just thin wrappers around reusable (business) logic stored under `pkg/rules/*` and `pkg/tracing/*` packages. 

### Overview 

Here are the main changes to the project structure: 

1. Only one Go module (instead of 3) at the root of the Git project to build tracee-ebpf and tracee-rules commands. (Makes life easier.) 
2. The `cmd/tracee-ebpf/main.go` is the entry point for the `tracee-ebpf` executable. It is supposed to be a client / wrapper of a logic provided by the `pkg/tracing/*` packages. 
3. The `cmd/tracee-rules/main.go` is the entry point for the `tracee-rules` executable. It is supposed to be a client / wrapper of a logic provided by the `pkg/rules/*` packages. 
4. The `pkg` folder is the root for Go code that is used by tracee-ebpf and tracee-rules executables, as well as any custom downstream integrations such as CNDR. The layout of the `pkg` folder must be determined but the main idea is to split Go code by functionality. For example, store code related to BPF programs under `pkg/tracing/*`, whereas code that’s related to signatures can be stored under `pkg/rules/*`.

### Example directory structure

<details>
<summary>Click to expand!</summary>

```
.
├── go.mod
├── go.sum
├── cmd
│   ├── tracee-ebpf
│   │   ├── main.go
│   │   └── main_test.go
│   └── tracee-rules
│       ├── input.go
│       ├── input_test.go
│       ├── main.go
│       ├── main_test.go
│       ├── output.go
│       ├── output_test.go
│       ├── signature.go
│       ├── signature_test.go
│       └── templates
│           ├── csv.tmpl
│           ├── falcosidekick.tmpl
│           ├── rawjson.tmpl
│           ├── simple.tmpl
│           ├── sprig.tmpl
│           └── xml.tmpl
├── pkg
│   ├── external
│   │   ├── external.go
│   │   └── external_test.go
│   └── rules
│       ├── engine
│       │   ├── engine.go
│       │   └── engine_test.go
│       ├── signatures
│       │   ├── golang
│       │   │   ├── examples
│       │   │   │   └── example.go
│       │   │   ├── export.go
│       │   │   ├── kubernetes_api_connection.go
│       │   │   ├── kubernetes_api_connection_test.go
│       │   │   ├── stdio_over_socket.go
│       │   │   └── stdio_over_socket_test.go
│       │   ├── helpers
│       │   │   └── helpers.go
│       │   ├── rego
│       │   │   └── regosig
│       │   │       ├── aio.go
│       │   │       ├── aio.rego
│       │   │       ├── aio_test.go
│       │   │       ├── common_test.go
│       │   │       ├── mapper.go
│       │   │       ├── mapper_test.go
│       │   │       ├── traceerego.go
│       │   │       └── traceerego_test.go
│       │   └── signaturestest
│       │       └── signaturestest.go
│       └── types
│           └── types.go
├── signatures
│   ├── anti_debugging_ptraceme.rego
│   ├── anti_debugging_ptraceme_test.rego
│   ├── cgroup_release_agent_modification.rego
│   ├── cgroup_release_agent_modification_test.rego
│   ├── code_injection.rego
│   ├── code_injection_test.rego
│   ├── disk_mount.rego
│   ├── disk_mount_test.rego
│   ├── dropped_executable.rego
│   ├── dropped_executable_test.rego
│   ├── dynamic_code_loading.rego
│   ├── dynamic_code_loading_test.rego
│   ├── embedded.go
│   ├── fileless_execution.rego
│   ├── fileless_execution_test.rego
│   ├── helpers.rego
│   ├── illegitimate_shell.rego
│   ├── illegitimate_shell_test.rego
│   ├── k8s_service_account_token.rego
│   ├── k8s_service_account_token_test.rego
│   ├── kernel_module_loading.rego
│   ├── kernel_module_loading_test.rego
│   ├── kubernetes_certificate_theft_attempt.rego
│   ├── kubernetes_certificate_theft_attempt_test.rego
│   ├── ld_preload.rego
│   └── ld_preload_test.rego
├── tracee-ebpf
│   ├── 3rdparty
│   │   ├── include
│   │   │   └── stdarg.h
│   │   └── libbpf
│   │       ├── BPF-CHECKPOINT-COMMIT
│   │       ├── CHECKPOINT-COMMIT
│   │       ├── LICENSE
│   │       ├── LICENSE.BSD-2-Clause
│   │       ├── LICENSE.LGPL-2.1
│   │       ├── README.md
│   │       ├── docs
│   │       │   ├── api.rst
│   │       │   ├── conf.py
│   │       │   ├── index.rst
│   │       │   ├── libbpf_build.rst
│   │       │   ├── libbpf_naming_convention.rst
│   │       │   └── sphinx
│   │       │       ├── Makefile
│   │       │       ├── doxygen
│   │       │       │   └── Doxyfile
│   │       │       └── requirements.txt
│   │       ├── include
│   │       │   ├── asm
│   │       │   │   └── barrier.h
│   │       │   ├── linux
│   │       │   │   ├── compiler.h
│   │       │   │   ├── err.h
│   │       │   │   ├── filter.h
│   │       │   │   ├── kernel.h
│   │       │   │   ├── list.h
│   │       │   │   ├── overflow.h
│   │       │   │   ├── ring_buffer.h
│   │       │   │   └── types.h
│   │       │   └── uapi
│   │       │       └── linux
│   │       │           ├── bpf.h
│   │       │           ├── bpf_common.h
│   │       │           ├── btf.h
│   │       │           ├── if_link.h
│   │       │           ├── if_xdp.h
│   │       │           ├── netlink.h
│   │       │           ├── pkt_cls.h
│   │       │           └── pkt_sched.h
│   │       ├── scripts
│   │       │   ├── coverity.sh
│   │       │   └── sync-kernel.sh
│   │       ├── src
│   │       │   ├── Makefile
│   │       │   ├── bpf.c
│   │       │   ├── bpf.h
│   │       │   ├── bpf_core_read.h
│   │       │   ├── bpf_endian.h
│   │       │   ├── bpf_gen_internal.h
│   │       │   ├── bpf_helper_defs.h
│   │       │   ├── bpf_helpers.h
│   │       │   ├── bpf_prog_linfo.c
│   │       │   ├── bpf_tracing.h
│   │       │   ├── btf.c
│   │       │   ├── btf.h
│   │       │   ├── btf_dump.c
│   │       │   ├── gen_loader.c
│   │       │   ├── hashmap.c
│   │       │   ├── hashmap.h
│   │       │   ├── libbpf.c
│   │       │   ├── libbpf.h
│   │       │   ├── libbpf.map
│   │       │   ├── libbpf.pc.template
│   │       │   ├── libbpf_common.h
│   │       │   ├── libbpf_errno.c
│   │       │   ├── libbpf_internal.h
│   │       │   ├── libbpf_legacy.h
│   │       │   ├── libbpf_probes.c
│   │       │   ├── linker.c
│   │       │   ├── netlink.c
│   │       │   ├── nlattr.c
│   │       │   ├── nlattr.h
│   │       │   ├── relo_core.c
│   │       │   ├── relo_core.h
│   │       │   ├── ringbuf.c
│   │       │   ├── skel_internal.h
│   │       │   ├── str_error.c
│   │       │   ├── str_error.h
│   │       │   ├── strset.c
│   │       │   ├── strset.h
│   │       │   ├── xsk.c
│   │       │   └── xsk.h
│   │       └── travis-ci
│   │           ├── managers
│   │           │   ├── debian.sh
│   │           │   ├── test_compile.sh
│   │           │   ├── travis_wait.bash
│   │           │   └── ubuntu.sh
│   │           └── vmtest
│   │               ├── build_pahole.sh
│   │               ├── build_selftests.sh
│   │               ├── checkout_latest_kernel.sh
│   │               ├── configs
│   │               │   ├── INDEX
│   │               │   ├── blacklist
│   │               │   │   ├── BLACKLIST-5.5.0
│   │               │   │   └── BLACKLIST-latest
│   │               │   ├── latest.config
│   │               │   └── whitelist
│   │               │       └── WHITELIST-4.9.0
│   │               ├── helpers.sh
│   │               ├── mkrootfs.sh
│   │               ├── prepare_selftests-4.9.0.sh
│   │               ├── prepare_selftests-5.5.0.sh
│   │               ├── prepare_selftests.sh
│   │               ├── run.sh
│   │               ├── run_selftests.sh
│   │               ├── run_vmtest.sh
│   │               └── vmlinux.h
│   ├── Dockerfile
│   ├── Dockerfile.builder
│   ├── Makefile
│   ├── Readme.md
│   ├── embedded.go
│   └── tracee
│       ├── argprinters.go
│       ├── argprinters_test.go
│       ├── buckets_cache.go
│       ├── co-re
│       │   ├── missing_definitions.h
│       │   ├── struct_flavors.h
│       │   └── vmlinux.h
│       ├── consts.go
│       ├── consts_amd64.go
│       ├── consts_arm64.go
│       ├── containers.go
│       ├── events.go
│       ├── events_decoder.go
│       ├── events_processor.go
│       ├── filters.go
│       ├── net_events.go
│       ├── printer.go
│       ├── tracee.bpf.c
│       ├── tracee.go
│       ├── tracee_test.go
│       ├── user_mode_events.go
│       ├── user_mode_events_test.go
│       └── write_capture.go
├── tracee-rules
│    ├── Dockerfile.builder
│    ├── Makefile
│    └── Readme.md
├── deploy
│   └── kubernetes
│       ├── tracee-falcosidekick
│       │   ├── falcosidekick.yaml
│       │   └── tracee.yaml
│       └── tracee-postee
│           └── tracee.yaml
├── Dockerfile
├── Makefile
├── Readme.md
├── Vagrantfile
└── entrypoint.sh
```
</details>

### Example downstream project (like CNDR) 

The following listings show the Go module definition and the entry point of a sample project (e.g., custom tracee-rules command) that imports the whole `github.com/aquasecurity/tracee` Go module, but uses only the `github.com/aquasecurity/tracee/pkg/rules/engine` and `github.com/aquasecurity/tracee/pkg/rules/types` packages.

```go
// go.mod
module github.com/aquasecurity/testdrive

go 1.16

require github.com/aquasecurity/tracee v0.6.4
```
```go
// cmd/testdrive/main.go
package main

import (
	"log"
	"os"

	"github.com/aquasecurity/tracee/pkg/rules/engine"
	"github.com/aquasecurity/tracee/pkg/rules/types"
)

func main() {
	// 1. Load signatures Rego and Go signatures
	signatures := make([]types.Signature, 0)
	// 2. Setup channel for input events
	sources := engine.EventSources{
		Tracee: make(chan types.Event),
	}
	// 3. Setup channel for output findings
	findingsCh := make(chan types.Finding)
	// 4. Construct rules engine
	engine, err := engine.NewEngine(signatures, sources, findingsCh, os.Stdout, true)
	if err != nil {
		log.Fatalf("constructing rules engine failed: %v", err)
	}
	done := make(chan bool)
	log.Println("Rules engine started and is waiting for events ...")
	// 5. Start rules engine
	engine.Start(done)
	<-done
}
```

The downstream project can be statically compiled with `CGO_ENABLED=0 go build -o dist/testdrive cmd/testdrive/main.go` (so long the upstream tracee project does not use `import "C"` in source code under `pkg/rules/*`).

## Other Considerations 

1. Put Rego and Go signatures somewhere in the top-level directory of tracee project (e.g., `./signatures`). Currently they are "hidden" deep in the hierarchy and mixed with Go code. 
2. Consider a "module" that groups source code and build artifacts to compile tracee kernel code (BPF objects) so it’s not mixed up with "pure" Go code.
3. Consider different kinds of tests and how to run them as part of the PR validation workflow. Some tests are not run at all. Sometimes we run integration tests along with unit tests (e.g. #1228).
4. A very experimental draft of what’s been described above #1198 - just to prove the concept. To refactor we should do agree on refactor steps and do it iteratively.

### Phase 1 (project layout, Make targets, CI workflow)

- [x] #1223 
- [x] #1220 
- [x] #1199
- [x] #1229 
- [x] #1233 
- [x] #1243 

### Phase 2 (C and Go code, integration tests, docs)

- [ ] #505 
- [ ] #960
- [ ] #1191 
- [ ] #1234 
- [ ] #1221 
- [ ] #1228

**Comment by simar7** [comment]
This looks good to me. A few points:

1. I agree that we should put signatures on the top level as it is an important first class citizen of Tracee. But instead of just doing them for Rego, we should include Go signatures as well at the top level. This probably introduces a little bit of complexity when compiling them but should at least keep them organized and structured in a single place, further reducing the coupling between Go signatures and Go code.
2. Client such as benchmarks and integrations tests, where would they go?

**Comment by yanivagman** [comment]
This is a good start, thanks @danielpacak
A few points:
1. We currently have "3rdparty" dir under tracee-ebpf - maybe we should move it to the top-level dir?
2. I agree with @simar7 that we should keep the go and rego signatures in a single place, as splitting them might be confusing.
3. Do we also want to have an "internal" dir for private packages that we don't want downstream projects to import? This can be used, for example, if we will decide to have a "decoder" package in tracee-ebpf
4. tracee-ebpf/tracee package can be moved, as a first step, to /pkg/tracing package. Later on we can break it into several packages (e.g. containers, filters, [ebpf]loader, events, network, capture)
5. tracee-ebpf and tracee-rules dirs should probably be removed

**Comment by danielpacak** [comment]
> This is a good start, thanks @danielpacak A few points:
> 
> 1. We currently have a `3rdparty` dir under tracee-ebpf - maybe we should move it to the top-level dir?
> 2. I agree with @simar7 that we should keep the go and rego signatures in a single place, as splitting them might be confusing.
> 3. Do we also want to have an `internal` dir for private packages that we don't want downstream projects to import? This can be used, for example, if we will decide to have a "decoder" package in tracee-ebpf that is only used internaly.
> 4. `tracee-ebpf/tracee` package can be moved, as a first step, to `/pkg/tracing` package. Later on we can break it into several packages (e.g. containers, filters, [ebpf]loader, events, network, capture)
> 5. `tracee-ebpf` and `tracee-rules` dirs should probably be removed

Certainly agree with all your points @yanivagman  and what @simar7 wrote about Go rules.

What I've described so far in scope of this issue is probably Phase 1 of restructuring, where we'd like do a relatively safe refactoring to have a single Go module and be able to compile the whole project and keep pure / reusable code under `./pkg/external`.

In Phase 2 we can work in parallel and further restructure individual modules such as refactor tracee-ebpf, tracee-rules, think about 3rdpary location, integration tests, etc.

**Comment by itaysk** [comment]
I'll start to create issues for the obvious packages, and we can discuss the other ones in a call

**Comment by itaysk** [comment]
@danielpacak do you think we should close this issue and track the rest in the milestone/in #960 ?

**Comment by danielpacak** [comment]
> @danielpacak do you think we should close this issue and track the rest in the milestone/in #960 ?

Yeah, I think it was useful to brainstorm, get started, and complete what we called "Phase 1". Phase 2 is a bunch of follow up tasks that can be tracked independently.
